iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergBufferedDeleteWriter.java

```diff
@@ -0,0 +1,179 @@
+  private final Map<PartitionKey, Map<String, Roaring64Bitmap>> buffer = Maps.newHashMap();
+  private final Map<Integer, PartitionSpec> specs;
+  private final Map<PartitionKey, PartitionSpec> keyToSpec = Maps.newHashMap();
+  private final FileFormat format;
+  private final FileWriterFactory<Record> writerFactory;
+  private final OutputFileFactory fileFactory;
+  private final FileIO io;
+  private final long targetFileSize;
+  private final Configuration configuration;
+  private final Record record;
+  private final InternalRecordWrapper wrapper;
+  private FilesForCommit filesForCommit;
+
+  HiveIcebergBufferedDeleteWriter(Schema schema, Map<Integer, PartitionSpec> specs, FileFormat format,
+      FileWriterFactory<Record> writerFactory, OutputFileFactory fileFactory, FileIO io, long targetFileSize,
+      Configuration configuration) {
+    this.specs = specs;
+    this.format = format;
+    this.writerFactory = writerFactory;
+    this.fileFactory = fileFactory;
+    this.io = io;
+    this.targetFileSize = targetFileSize;
+    this.configuration = configuration;
+    this.wrapper = new InternalRecordWrapper(schema.asStruct());
+    this.record = GenericRecord.create(schema);
+  }
+
+  @Override
+  public void write(Writable row) throws IOException {
+    Record rec = ((Container<Record>) row).get();
+    IcebergAcidUtil.getOriginalFromUpdatedRecord(rec, record);
+    String filePath = (String) rec.getField(MetadataColumns.FILE_PATH.name());
+    int specId = IcebergAcidUtil.parseSpecId(rec);
+
+    Map<String, Roaring64Bitmap> deleteMap =
+        buffer.computeIfAbsent(partition(record, specId), key -> {
+          keyToSpec.put(key, specs.get(specId));
+          return Maps.newHashMap();
+        });
+    Roaring64Bitmap deletes = deleteMap.computeIfAbsent(filePath, unused -> new Roaring64Bitmap());
+    deletes.add((Long) rec.getField(MetadataColumns.ROW_POSITION.name()));
+  }
+
+  @Override
+  public void close(boolean abort) throws IOException {
+    long startTime = System.currentTimeMillis();
+    Collection<DeleteFile> deleteFiles = new ConcurrentLinkedQueue<>();
+    if (!abort) {
+      LOG.info("Delete file flush is started");
+      ExecutorService fileExecutor = fileExecutor(configuration, buffer.size());
+      try {
+        Tasks.foreach(buffer.keySet())
+            .retry(3)
+            .executeWith(fileExecutor)
+            .onFailure((partition, exception) -> LOG.info("Failed to write delete file {}", partition, exception))
+            .run(partition -> {
+              PositionDelete<Record> positionDelete = PositionDelete.create();
+              PartitioningWriter writerForFiles;
+              try (PartitioningWriter writer =
+                       new ClusteredPositionDeleteWriter<>(writerFactory, fileFactory, io, format, targetFileSize)) {
+                Map<String, Roaring64Bitmap> partitionData = buffer.get(partition);
+                for (String filePath : new TreeSet<>(partitionData.keySet())) {
```

Comment:
Would it make sense to store the filepaths already in a TreeMap? then this step could be skipped

Reply:
https://stackabuse.com/hashmap-and-treemap-in-java-differences-and-similarities/

> TreeMap is based on binary tree that provides time performance O(log(n)).
> Thus, HashMap almost always works faster than TreeMap. The larger the object that's stored, the faster HashMap will be in comparison to TreeMap. However, a TreeMap uses the optimal amount of memory to hold its items, unlike a HashMap.

I was aiming for `write` method performance, and decided to focus on that instead of the `close` performance.

Reply:
I see. My thinking was that this takes `new TreeSet<>(partitionData.keySet())` n*logn to sort the data, so overall the logn insertion for the TreeMap will still be cheaper.

Reply:
My fear is that TreeSet.get() is log(n) too. And this is what we do for every record, but I might be wrong

Reply:
Yes, I think you're right. Anyways, I have no strong feelings on this, just wanted to discuss it. We can leave it as is.